// splunk-ui/src/components/Table/Cell.jsx
import React, { Component, PropTypes } from 'react';
import { keycode } from 'splunk-ui/util/keyboard';
import toClassName from 'splunk-ui/util/toClassName';
import css from './Cell.css';

function noop() {}


class Cell extends Component {
    static splunkUiType = 'Table.Cell';

    static propTypes = {
        /** @docs-ignore. */
        children: PropTypes.node,
        /** @docs-ignore. */
        className: PropTypes.string,
        /** @docs-ignore. This will be passed through, and will work as expected. */
        onClick: PropTypes.func,
        /** @docs-ignore. This will be passed through, and will work as expected. */
        onKeyDown: PropTypes.func,
    };

    static defaultProps = {
        onClick: noop,
        onKeyDown: noop,
    };

    handleKeyDown = (e) => {
        const { onClick, onKeyDown } = this.props;
        if (keycode(e) === 'enter') {
            onClick(e);
        }
        onKeyDown(e);
    }

    render() {
        const {
            children,
            className,
            onKeyDown, // eslint-disable-line no-unused-vars
            ...other
        } = this.props;
        return (
            <td
                className={toClassName(css.main, className)}
                onKeyDown={this.handleKeyDown}
                {...other}
            >
                {children}
            </td>
        );
    }
}

export default Cell;
