// splunk-ui/src/components/TabBar/TabBar.jsx
import React, { PropTypes } from 'react';
import { omit } from 'lodash';
import { createTestHook } from 'splunk-ui/util/testSupport';
import toClassName from 'splunk-ui/util/toClassName';
import Tab from './Tab';
import css from './TabBar.css';

function noop() {}

const propTypes = {
    /**
     * `children` should be `TabBar.Tab`.
     **/
    children: PropTypes.node,
    /** @docs-ignore */
    className: PropTypes.string,
    /** A callback that receives the event and data (selectedTabId). */
    onChange: PropTypes.func.isRequired,
    /** The `tabId` of the `TabBar.Tab` to activate. */
    activeTabId: PropTypes.any,
};

const defaultProps = {
    onChange: noop,
};

function TabBar(props) {
    const { activeTabId, className, children } = props;
    const classes = toClassName(css.main, className);
    const clonedChildren = React.Children.map(children, (child) =>
        React.cloneElement(child, {
            onClick: (e, data) => {
                props.onChange(e,
                    { selectedTabId: data.tabId }
                );
            },
            active: activeTabId === child.props.tabId,
        })
    );
    return (
        <ul
            className={classes}
            {...createTestHook(__filename)}
            {...omit(props, Object.keys(propTypes))}
        >
            {clonedChildren}
        </ul>
    );
}

TabBar.propTypes = propTypes;
TabBar.defaultProps = defaultProps;
TabBar.Tab = Tab;

export default TabBar;
