// splunk-ui/src/components/File/examples/CustomChildren.jsx
import React, { Component } from 'react';
import File from 'splunk-ui/components/File';
import X from 'splunk-ui/icons/X';
import Document from 'splunk-ui/icons/Document';

const itemStyles = {
    background: 'green',
    color: 'white',
    padding: 10,
    position: 'relative',
};
const docStyles = {
    fontSize: 80,
    opacity: 0.2,
    marginRight: 10,
};
const xStyles = {
    position: 'absolute',
    right: 10,
    top: 10,
};

class Example extends Component {
    constructor(props, ...rest) {
        super(props, ...rest);

        this.state = {};

        this.fileReader = new FileReader();
    }

    handleAddFiles = (files) => {
        const file = files[0];

        if (this.fileReader.readyState === 1) {
            this.fileReader.abort();
        }

        this.fileReader.onload = function onLoad() {
            this.setState({
                filename: file.name,
                value: this.fileReader.result,
            });
        }.bind(this);

        this.fileReader.readAsBinaryString(file);
    }

    handleRemoveFile = () => {
        if (this.fileReader.readyState === 1) {
            this.fileReader.abort();
        }

        this.setState({
            filename: null,
            value: null,
        });
    }

    render() {
        return (
            <div style={{ maxWidth: 300 }}>
                <File
                    onAddFiles={this.handleAddFiles}
                    onRemoveFile={this.handleRemoveFile}
                >
                    {this.state.filename && (
                        <div style={itemStyles}>
                            <Document style={docStyles} />
                            {this.state.filename}
                            <X style={xStyles} />
                        </div>
                    )}
                </File>
            </div>
        );
    }
}

export default Example;
