// splunk-ui/src/components/File/examples/Progress.jsx
import React, { Component } from 'react';
import File from 'splunk-ui/components/File';


class Example extends Component {
    constructor(props, ...rest) {
        super(props, ...rest);

        this.state = {};

        this.fileReader = new FileReader();
    }

    handleAddFiles = (files) => {
        const file = files[0];

        if (this.fileReader.readyState === 1) {
            this.fileReader.abort();
        }

        this.fileReader.onload = () => {
            this.setState({
                filename: file.name,
                value: this.fileReader.result,
                uploadPercentage: 0,
            });

            this.fakeUpload();
        };

        this.fileReader.readAsBinaryString(file);
    }

    handleRemoveFile = () => {
        if (this.fileReader.readyState === 1) {
            this.fileReader.abort();
        }

        this.setState({
            filename: null,
            value: null,
        });
    }

    fakeUpload = () => {
        if (this.state.uploadPercentage === 100) {
            this.setState({
                uploadPercentage: undefined,
            });

            delete this.timer;
            return;
        }

        if (!this.timer) {
            // show the queued state for 2 seconds.
            this.timer = setTimeout(this.fakeUpload, 2000);
        } else {
            this.setState({
                uploadPercentage: this.state.uploadPercentage + 5,
            });

            this.timer = setTimeout(this.fakeUpload, 500);
        }
    }

    render() {
        return (
            <div style={{ maxWidth: 300 }}>
                <File
                    size="medium"
                    help="this.handleAddFiles = this.handleAddFiles.bind(this);"
                    onAddFiles={this.handleAddFiles}
                    onRemoveFile={this.handleRemoveFile}
                >

                    {this.state.filename && (
                        <File.Item
                            name={this.state.filename}
                            uploadPercentage={this.state.uploadPercentage}
                        />
                    )}
                </File>
            </div>
        );
    }
}

export default Example;
