// splunk-ui/src/components/File/examples/Single.jsx
import React, { Component } from 'react';
import File from 'splunk-ui/components/File';

class Example extends Component {
    constructor(props, ...rest) {
        super(props, ...rest);

        this.state = {};

        this.fileReader = new FileReader();
    }

    handleAddFiles = (files) => {
        const file = files[0];

        if (this.fileReader.readyState === 1) {
            this.fileReader.abort();
        }

        this.fileReader.onload = () => {
            this.setState({
                filename: file.name,
                value: this.fileReader.result,
            });
        };

        this.fileReader.readAsBinaryString(file);
    }

    handleRemoveFile = () => {
        if (this.fileReader.readyState === 1) {
            this.fileReader.abort();
        }

        this.setState({
            filename: null,
            value: null,
        });
    }

    render() {
        return (
            <div style={{ maxWidth: 300 }}>
                <File
                    help="this.handleAddFiles = this.handleAddFiles.bind(this);"
                    onAddFiles={this.handleAddFiles}
                    onRemoveFile={this.handleRemoveFile}
                >
                    {this.state.filename && <File.Item name={this.state.filename} />}
                </File>
            </div>
        );
    }
}

export default Example;
