// splunk-ui/tools/templates/statefulComponent.js
/* eslint-disable max-len */

import { extend } from 'lodash';
import baseComponent from './baseComponent';

export default function statefulComponent(options) {
    const { componentName } = options;
    return extend({
        // Stub of the main JavaScript file.
        [`src/components/${componentName}/${componentName}.jsx`]: `import React, { Component, PropTypes } from 'react';
import { createTestHook } from 'splunk-ui/util/testSupport';
import toClassName from 'splunk-ui/util/toClassName';
import css from './${componentName}.css';

class ${componentName} extends Component {
    static propTypes = {
        className: PropTypes.string,
        name: PropTypes.string.isRequired,
    };

    static defaultProps = {
        name: 'Joe',
    };

    constructor(props, ...rest) {
        super(props, ...rest);
        this.state = {
            hasState: props.hasState,
        };
    }

    render() {
        const { className, name, ...otherProps } = this.props;
        const message = this.state.hasState ? 'I have state' : 'I am stateless';
        return (
            <div
                className={toClassName(css.main, className)}
                {...createTestHook(__filename)}
                {...otherProps}
            >
                Hello {name}, {message}
            </div>
        );
    }
}

export default ${componentName};
`,
    }, baseComponent(options));
}
