// splunk-ui/docs/webpack.config.js
var path = require('path');
var fs = require('fs');
var extend = require('extend');
var webpack = require('webpack');
var CopyWebpackPlugin = require('copy-webpack-plugin');
var HtmlWebpackPlugin = require('html-webpack-plugin');
var baseConfig = require('../tools/base.config');

var DEBUG = process.env.NODE_ENV !== 'production';

module.exports = extend(true, {}, baseConfig, {
    entry: 'index',
    resolve: {
        root: path.join(__dirname, 'src'),
        alias: {
            'splunk-ui': path.join(__dirname, '../src'),
        },
    },
    output: {
        path: path.join(__dirname, 'build'),
        filename: '[name].js',
    },
    plugins: [
        new CopyWebpackPlugin([
            { from: path.join(__dirname, 'assets') }
        ]),
        new HtmlWebpackPlugin({
            hash: true,
            template: path.join(__dirname, 'src/index.html'),
        }),
        // Define free variables
        // https://webpack.github.io/docs/list-of-plugins.html#defineplugin
        new webpack.DefinePlugin({
            'process.env.NODE_ENV': DEBUG ? '"development"' : '"production"',
            __DEV__: DEBUG,
        }),
    ].concat(baseConfig.plugins).concat(DEBUG ? [] : [
        // Search for equal or similar files and deduplicate them in the output
        // https://webpack.github.io/docs/list-of-plugins.html#dedupeplugin
        new webpack.optimize.DedupePlugin(),

        // Minimize all JavaScript output of chunks
        // https://github.com/mishoo/UglifyJS2#compressor-options
        new webpack.optimize.UglifyJsPlugin({
            compress: {
                screw_ie8: true,
                warnings: false,
            },
        }),

        // A plugin for a more aggressive chunk merging strategy
        // https://webpack.github.io/docs/list-of-plugins.html#aggressivemergingplugin
        new webpack.optimize.AggressiveMergingPlugin(),
    ]),
    devtool: DEBUG ? 'eval-source-map' : false,
});
