// splunk-ui/src/components/Logo/examples/Dark.jsx
import React from 'react';
import { Splunk, Enterprise, Light, Cloud, Hunk } from 'splunk-ui/components/Logo';

function DarkExample() {
    return (
        <div
            style={{
                display: 'flex',
                flexWrap: 'wrap',
                fontSize: '2em',
                backgroundColor: '#333',
                padding: '30px 0',
            }}
        >
            <div style={{ flexBasis: '50%', padding: '30px 60px', boxSizing: 'border-box' }}>
                <Splunk invert />
            </div>
            <div style={{ flexBasis: '50%', padding: '30px 60px', boxSizing: 'border-box' }}>
                <Enterprise invert />
            </div>
            <div style={{ flexBasis: '50%', padding: '30px 60px', boxSizing: 'border-box' }}>
                <Light invert />
            </div>
            <div style={{ flexBasis: '50%', padding: '30px 60px', boxSizing: 'border-box' }}>
                <Cloud invert />
            </div>
            <div style={{ flexBasis: '50%', padding: '30px 60px', boxSizing: 'border-box' }}>
                <Hunk invert />
            </div>
        </div>
    );
}

export default DarkExample;
