// splunk-ui/src/components/Message/Message.jsx
import React, { PropTypes } from 'react';
import { createTestHook } from 'splunk-ui/util/testSupport';
import toClassName from 'splunk-ui/util/toClassName';
import Box from 'splunk-ui/components/Box';
import AlertSquare from 'splunk-ui/icons/AlertSquare';
import AlertTriangle from 'splunk-ui/icons/AlertTriangle';
import CheckCircle from 'splunk-ui/icons/CheckCircle';
import InfoCircle from 'splunk-ui/icons/InfoCircle';
import css from './Message.css';

const propTypes = {
    /** @docs-ignore */
    className: PropTypes.string,
    /** Content of the message */
    children: PropTypes.node,
    /** The severity or type of warning */
    type: PropTypes.oneOf(['info', 'success', 'warning', 'error']),
};

const defaultProps = {
    type: 'warning',
};

export default function Message(props) {
    const { type, children, className, ...otherProps } = props;
    const Icon = {
        info: InfoCircle,
        success: CheckCircle,
        warning: AlertSquare,
        error: AlertTriangle,
    }[type];

    return (
        <Box
            className={toClassName(css.main, className)}
            {...createTestHook(__filename)}
            {...otherProps}
        >
            <span className={css[type]}>
                <Icon size={1.6} />
            </span>
            {children}
        </Box>
    );
}

Message.propTypes = propTypes;
Message.defaultProps = defaultProps;
