// splunk-ui/src/components/Scroll/examples/Controlled.jsx
import React, { Component } from 'react';
import Button from 'splunk-ui/components/Button';
import Scroll from 'splunk-ui/components/Scroll';

class Basic extends Component {

    handleMount = (scrollComp) => {
        this.setState({ scrollComp });
    }


    handleClickDown = () => {
        const c = this.state.scrollComp;

        c.scrollTo({ top: c.scrollTop() + 200 });
    }

    handleClickUp = () => {
        const c = this.state.scrollComp;

        c.scrollTo({ top: c.scrollTop() - 200 });
    }

    render() {
        const styles = {
            border: '1px solid #ccc',
            height: '400px',
            width: '400px',
            marginTop: '20px',
        };

        return (
            <div>
                <Button label="Scroll Down" onClick={this.handleClickDown} />
                <Button label="Scroll Up" onClick={this.handleClickUp} />

                <Scroll
                    ref={this.handleMount}
                    style={styles}
                >
                    { new Array(400).join('Repeating words. ') }
                </Scroll>
            </div>
        );
    }
}

export default Basic;
