// splunk-ui/src/components/Modal/examples/Undismissable.jsx
import React, { Component } from 'react';
import Button from 'splunk-ui/components/Button';
import Modal from 'splunk-ui/components/Modal';

class Basic extends Component {
    constructor(props, context) {
        super(props, context);
        this.state = {
            open: false,
            count: null,
        };
    }
    componentWillUnmount() {
        window.clearInterval(this.counterInterval);
    }

    handleCounter = () => {
        const newCount = this.state.count - 1;
        if (!newCount) {
            this.setState({
                open: false,
            });
            window.clearInterval(this.counterInterval);
        } else {
            this.setState({
                count: newCount,
            });
        }
    }

    handleRequestOpen = () => {
        this.counterInterval = window.setInterval(this.handleCounter, 1000);
        this.setState({
            count: 5,
            open: true,
        });
    }

    render() {
        const { open, count } = this.state;
        return (
            <div>
                <Button onClick={this.handleRequestOpen} label="Click me" />
                <Modal open={open}>
                    <Modal.Body>
                        This modal cannot be dismissed by the user. It will be
                        dismissed in {count} seconds.
                    </Modal.Body>
                </Modal>
            </div>
        );
    }
}

export default Basic;
