// splunk-ui/src/components/TransitionOpen/examples/ExpandWidth.jsx
import React, { Component } from 'react';
import Button from 'splunk-ui/components/Button';
import TransitionOpen from 'splunk-ui/components/TransitionOpen';

class ExpandWidth extends Component {
    constructor(props, context) {
        super(props, context);
        this.state = {
            open: false,
        };
    }

    handleClick = () => {
        this.setState({ open: !this.state.open });
    }

    render() {
        const style = {
            border: '1px solid black',
            marginTop: 10,
            textAlign: 'center',
            width: 150,
        };
        return (
            <div style={{ height: 200 }}>
                <Button onClick={this.handleClick}>Click Me!</Button>
                <TransitionOpen animation="expandWidth" open={this.state.open}>
                    <div style={style}>
                        <p>Hello World</p>
                        <p>Hello World</p>
                        <p>Hello World</p>
                        <p>Hello World</p>
                        <p>Hello World</p>
                    </div>
                </TransitionOpen>
            </div>
        );
    }
}

export default ExpandWidth;
