// splunk-ui/src/components/Menu/Item.jsx
import React, { Component, PropTypes } from 'react';
import { _ } from 'splunk-ui/util/i18n';
import { createTestHook } from 'splunk-ui/util/testSupport';
import toClassName from 'splunk-ui/util/toClassName';
import Clickable from 'splunk-ui/components/Clickable';
import Check from 'splunk-ui/icons/Check';
import css from './Item.css';

class Item extends Component {
    static propTypes = {
        /**
         * Active shows a temporary selected state similar to focus. This is used when filtering the
         * menu items in Multiselect, Select and ComboBox and navigating with arrows.
         */
        active: PropTypes.bool,
        /** @docs-ignore */
        className: PropTypes.string,
        /**
        * Additional information to explain the option, such as "Recommended".
        */
        description: PropTypes.string,
        /**
        * The description text may appear to the right of the label or under the label.
        */
        descriptionPosition: React.PropTypes.oneOf(['right', 'bottom']),
        /**
        * If disabled=true, the item is grayed out and cannot be clicked.
        */
        disabled: PropTypes.bool,
        /**
        * An icon to insert before the label.
        */
        icon: PropTypes.node,
        /**
        * The text to show for the item.
        */
        label: PropTypes.string.isRequired,
        /**
        * Callback for click events.
        */
        onClick: PropTypes.func,
        /**
        * To open the link in a new window, set openInNewContext to true. An icon will be added
        * indicating the the behavior
        */
        openInNewContext: PropTypes.bool,
        /**
        * If selectable is true, whitespace will be left where the checkmark can be shown.
        */
        selectable: PropTypes.bool,
        /**
        * If selected is true, a checkmark will show the item is selected.
        */
        selected: PropTypes.bool,
        /* A url or path to link to.  */
        to: PropTypes.string,
        /**
        * If truncate=false, wrapping will be disabled and any additional text will be ellipsised.
        */
        truncate: PropTypes.bool,
    }

    static defaultProps = {
        active: false,
        descriptionPosition: 'bottom',
        disabled: false,
        openInNewContext: false,
        selectable: false,
        selected: false,
        truncate: false,
    }

    scrollIntoViewIfNeeded() {
        const parentEl = this.el.offsetParent;
        if (!parentEl) return;

        // Below the bottom of the container.
        if (parentEl.scrollTop + parentEl.clientHeight < this.el.offsetTop + this.el.clientHeight) {
            parentEl.scrollTop = (this.el.offsetTop + this.el.clientHeight) - parentEl.clientHeight;

        // Above the top of the container.
        } else if (parentEl.scrollTop > this.el.offsetTop) {
            parentEl.scrollTop = this.el.offsetTop;
        }
    }

    render() {
        const {
            active,
            className,
            selectable,
            selected,
            icon,
            description,
            disabled,
            onClick,
            to,
            truncate,
            descriptionPosition,
            openInNewContext,
            ...other
        } = this.props;

        return (
            <li
                className={css.item}
                ref={el => this.el = el}
                {...createTestHook(__filename)}
                {...other}
            >
                <Clickable
                    className={toClassName(css.link, className)}
                    data-selectable={selectable || selected}
                    data-selected={selected}
                    data-has-icon={!!icon}
                    data-truncation={truncate}
                    data-active={active ? true : null}
                    disabled={disabled}
                    onClick={onClick}
                    to={to}
                    title={truncate ? this.props.label : null}
                    openInNewContext={openInNewContext}
                >
                    {selected && (
                        <div className={css.itemSelectedIcon}>
                            <Check screenReaderText={_('Selected')} size={0.85} />
                        </div>
                    )}
                    {icon && (
                        <span className={css.itemIcon}>
                            {icon}
                        </span>
                    )}
                    {description && descriptionPosition === 'right' && (
                        <span className={css.itemDescriptionRight}>
                            {description}
                        </span>
                    )}
                    {this.props.label}
                    {description && descriptionPosition === 'bottom' && (
                        <span className={css.itemDescriptionBottom}>
                            {description}
                        </span>
                    )}
                </Clickable>
            </li>
        );
    }
}
export default Item;
