// splunk-ui/src/components/Menu/Menu.jsx
import React, { PropTypes, Component } from 'react';
import { findDOMNode } from 'react-dom';
import { getSortedTabableElements, handleTab } from 'splunk-ui/util/focus';
import { keycode } from 'splunk-ui/util/keyboard';
import { createTestHook } from 'splunk-ui/util/testSupport';
import toClassName from 'splunk-ui/util/toClassName';
import Scroll from 'splunk-ui/components/Scroll';
import Item from './Item';
import Divider from './Divider';
import css from './Menu.css';

/**
* Menus are generally placed inside a Dropdown.
*/
class Menu extends Component {
    static propTypes = {
        /**
         * `children` should be `Menu.Item` or `Menu.Divider`.
         **/
        children: PropTypes.node,
        /** @docs-ignore */
        className: PropTypes.string,
        /** Keep focus within the menu while navigating by keyboard. Tabbing from the last item
         * returns user to the first item. */
        retainFocus: PropTypes.bool,
        useSyntheticScroll: PropTypes.bool,
    };

    static defaultProps = {
        retainFocus: true,
        useSyntheticScroll: false,
    };

    static Item = Item;
    static Divider = Divider;

    constructor(props, ...rest) {
        super(props, ...rest);

        this.state = {};

        this.handleMount = this.handleMount.bind(this);
        this.handleKeyDown = this.handleKeyDown.bind(this);
    }

    getEl() {
        return this.state.containerEl;
    }

    handleMount(scrollContainer) {
        // TODO: consider adding methods to Scroll for this to avoid using findDOMNode.
        this.setState({ containerEl: findDOMNode(scrollContainer) });
    }

    handleKeyDown(e) {
        if (keycode(e) === 'tab' && this.props.retainFocus) {
            handleTab(this.state.containerEl, e);
            return;
        }

        if (keycode(e) !== 'down' && keycode(e) !== 'up') {
            return;
        }

        const tabbableElements = getSortedTabableElements(this.state.containerEl);
        const currentIndex = tabbableElements.indexOf(e.target);

        if (currentIndex === -1) {
            return;
        }

        if (keycode(e) === 'up' && currentIndex > 0) {
            tabbableElements[currentIndex - 1].focus();
        } else if (keycode(e) === 'down' && currentIndex < tabbableElements.length - 1) {
            tabbableElements[currentIndex + 1].focus();
        }
        e.preventDefault();
    }

    render() {
        const {
            children,
            className,
            retainFocus, // eslint-disable-line no-unused-vars
            useSyntheticScroll,
            ...otherProps
        } = this.props;

        const ComponentType = useSyntheticScroll ? Scroll : 'ul';
        const syntheticScrollProps = useSyntheticScroll ? {
            component: 'ul',
            stopScrollPropagation: true,
        } : {};

        return (
            <ComponentType
                className={toClassName(css.main, className)}
                ref={this.handleMount}
                onKeyDown={this.handleKeyDown}
                {...syntheticScrollProps}
                {...createTestHook(__filename)}
                {...otherProps}
            >
                {children}
            </ComponentType>
        );
    }

}

export default Menu;
