// splunk-ui/src/components/ButtonGroup/ButtonGroup.jsx
import React, { Children, PropTypes } from 'react';
import { extend } from 'lodash';
import { createTestHook } from 'splunk-ui/util/testSupport';
import toClassName from 'splunk-ui/util/toClassName';
import Box from 'splunk-ui/components/Box';
import css from './ButtonGroup.css';

const propTypes = {
    /** @docs-ignore */
    children: PropTypes.node,
    /** @docs-ignore */
    className: PropTypes.string,
    /** @docs-ignore */
    inline: PropTypes.bool,
};

const defaultProps = {
    inline: true,
};

export default function ButtonGroup(props) {
    const { children, className, ...otherProps } = props;

    const childrenCleaned = Children.map(children, (item, i) => React.cloneElement(item, {
        prepend: i > 0,
        append: i < React.Children.count(children) - 1,
        key: item.key || i,
        inline: false,
        style: extend({ width: 'auto' }, item.style),
    }));

    return (
        <Box
            flex
            className={toClassName(css.main, className)}
            {...createTestHook(__filename)}
            {...otherProps}
        >
            {childrenCleaned}
        </Box>
    );
}

ButtonGroup.propTypes = propTypes;
ButtonGroup.defaultProps = defaultProps;
