// splunk-ui/src/components/Tooltip/examples/Controlled.jsx
import React, { Component } from 'react';
import Tooltip from 'splunk-ui/components/Tooltip';
import Text from 'splunk-ui/components/Text';


class Example extends Component {
    constructor(props, ...rest) {
        super(props, ...rest);

        this.state = {
            open: false,
            reason: 'none',
        };
    }

    componentWillUnmount() {
        clearTimeout(this.timeout);
    }

    handleRequestOpen = (e, { reason }) => {
        this.setState({ reason });
        clearTimeout(this.timeout);
        this.timeout = setTimeout(() => {
            this.setState({ open: true });
        }, 500);
    }

    handleRequestClosed = (e, { reason }) => {
        this.setState({ reason });
        clearTimeout(this.timeout);
        this.timeout = setTimeout(() => {
            this.setState({ open: false });
        }, 500);
    }

    render() {
        return (
            <div>
                <Tooltip
                    content="I explain the ambiguous control."
                    open={this.state.open}
                    onRequestOpen={this.handleRequestOpen}
                    onRequestClose={this.handleRequestClosed}
                >
                    <Text />
                </Tooltip><br />
                Received request reason: {this.state.reason}
            </div>
        );
    }
}

export default Example;
