// splunk-ui/src/components/Slider/Slider.jsx
import React, { Component, PropTypes } from 'react';
import toClassName from 'splunk-ui/util/toClassName';
import EventListener from 'react-event-listener';
import { createTestHook } from 'splunk-ui/util/testSupport';
import { has, noop, omit } from 'lodash';
import { invariant } from 'splunk-ui/util/errorHandling';
import { keycode } from 'splunk-ui/util/keyboard';
import Tooltip from 'splunk-ui/components/Tooltip';
import Box from 'splunk-ui/components/Box';
import css from './Slider.css';

class Slider extends Component {

    static propTypes = {
        /** @docs-ignore */
        className: PropTypes.string,
        /** Set this property instead of value to make value uncontrolled. */
        defaultValue: PropTypes.number,
        /** When false, display as inline-block with the default width. */
        inline: PropTypes.bool,
        /** The minimum value of the Slider input. */
        min: PropTypes.number,
        /** The maximum value of the Slider input. */
        max: PropTypes.number,
        /** Return event and data object with slider value */
        onChange: PropTypes.func,
        /** The step value of the Slider input. */
        step: PropTypes.number,
        /** The value of the slider.
        * Setting this value makes the property controlled. A callback is required. */
        value: PropTypes.number,
    };

    static defaultProps = {
        inline: false,
        min: 1,
        max: 5,
        onChange: noop,
        step: 1,
    };

    constructor(props, ...rest) {
        super(props, ...rest);

        this.controlledExternally = has(props, 'value');

        this.selected = false;
        this.state = {
            position: 50,
            showTooltip: false,
            value: this.isControlled() ? null : props.defaultValue,
        };

        if (__DEV__ && this.isControlled() && has(props, 'defaultValue')) {
            invariant('defaultValue prop is not compatible with value prop.');
        }

        if (props.min >= props.max && __DEV__) {
            invariant('Error in Slider: max must be greater than min');
        }
    }

    componentDidMount() {
        this.width = this.sliderBar.clientWidth;
        this.thumbRadius = this.sliderThumb.clientWidth / 2;
        /* eslint react/no-did-mount-set-state: 0 */
        this.setState({
            position: this.valueToPosition(this.getValue()),
        });
    }

    componentWillReceiveProps(nextProps) {
        if (__DEV__ && !this.isControlled() && has(nextProps, 'value')) {
            invariant(`Cannot change Slider from an uncontrolled component to a controlled
    one. Prop 'value' is not valid in subsequent updates if not provided in the initial props.`);
        }

        if (__DEV__ && nextProps.defaultValue !== this.props.defaultValue) {
            invariant('Cannot change defaultValue after set.');
        }
        this.setState({
            position: this.valueToPosition(nextProps.value),
        });
    }

    getValue() {
        return this.isControlled() ? this.props.value : this.state.value;
    }

    setValue(e, value) {
        this.setState({
            value: this.isControlled() ? null : value,
        }, () => {
            if (this.getValue() !== value) {
                this.props.onChange(e, { value });
            }
        });
    }

    isControlled() {
        return this.controlledExternally;
    }

    checkPositionBounds = (pos) => {
        if (pos > 100) {
            return 100;
        } else if (pos < 0) {
            return 0;
        }
        return pos;
    }

    checkValueBounds = (val) => {
        if (val > this.props.max) {
            return this.props.max;
        } else if (val < this.props.min) {
            return this.props.min;
        }
        return val;
    }

    handleBlur = () => {
        this.setState({
            showTooltip: false,
        });
    }

    handleClick = (e) => {
        const pos = this.percentageFromEvent(e);
        const val = this.roundValue(this.positionToValue(pos));
        this.setValue(e, val);

        this.setState({
            position: pos,
        }, this.snapToPosition());
    }

    handleFocus = () => {
        this.setState({
            showTooltip: true,
        });
    }

    handleKeyDown = (e) => {
        const step = this.props.step;
        let val = this.getValue();
        if (keycode(e) === 'right') {
            val = this.checkValueBounds(val + step);
        } else if (keycode(e) === 'left') {
            val = this.checkValueBounds(val - step);
        } else {
            return;
        }
        val = this.roundValue(val);
        this.setValue(e, val);
        this.setState({
            position: this.valueToPosition(this.roundValue(val)),
        });
    }

    handleMouseDown = (e) => {
        e.preventDefault();
        this.sliderThumb.focus();
        this.selected = true;
        this.setState({
            showTooltip: true,
        });
    }

    handleMouseEnter = () => {
        this.setState({
            showTooltip: true,
        });
    }

    handleMouseLeave = () => {
        if (!this.selected) {
            this.setState({
                showTooltip: false,
            });
        }
    }

    handleMouseMove = (e) => {
        if (this.selected) {
            const pos = this.percentageFromEvent(e);

            this.setValue(e, this.roundValue(this.positionToValue(pos)));
            this.setState({
                position: pos,
            }, this.props.onChange(e, { value: this.getValue() }));
        }
    }

    handleMouseUp = () => {
        this.selected = false;
        this.setState({
            showTooltip: false,
        }, this.snapToPosition);
        this.sliderThumb.blur();
    }

    percentageFromEvent = (e) => {
        const offset = e.clientX - this.sliderBar.getBoundingClientRect().left;
        return this.checkPositionBounds((offset / this.width) * 100);
    }

    positionToValue = (pos) => {
        const valRange = this.props.max - this.props.min;
        return ((pos / 100) * valRange) + this.props.min;
    }

    roundValue = (val) => {
        const step = this.props.step;
        const rem = val % step;
        if (rem < step / 2) {
            return val - rem;
        } else if (rem > step / 2) {
            return val + (step - rem);
        }
        return val;
    }

    snapToPosition = () => {
        this.setState({
            position: this.valueToPosition(this.getValue()),
        });
    }


    valueToPosition = (val) => {
        const valRange = this.props.max - this.props.min;
        return ((val - this.props.min) / valRange) * 100;
    }

    render() {
        const { className, min, max, ...otherProps } = this.props;
        const boxProps = omit(otherProps, ['step', 'value']);
        return (
            <Box
                className={toClassName(css.main, className)}
                flex
                {...createTestHook(__filename)}
                {...boxProps}
            >
                <EventListener
                    onMouseUp={this.handleMouseUp}
                    onMouseMove={this.handleMouseMove}
                    onResize={this.handleResize}
                    target="window"
                />
                <Box className={css.minLabel}>
                    {min}
                </Box>
                <div
                    className={css.input}
                    data-attribute="sliderBarWrapper"
                    onClick={this.handleClick}
                >
                    <div
                        className={css.sliderBar}
                        ref={el => this.sliderBar = el}
                    />
                    <Tooltip
                        className={css.tooltip}
                        content={this.getValue()}
                        open={this.state.showTooltip}
                        style={{ left: `${this.state.position}%` }}
                    >
                        <div
                            className={css.sliderThumb}
                            onBlur={this.handleBlur}
                            onMouseEnter={this.handleMouseEnter}
                            onMouseLeave={this.handleMouseLeave}
                            onKeyDown={this.handleKeyDown}
                            onMouseDown={this.handleMouseDown}
                            onFocus={this.handleFocus}
                            ref={el => this.sliderThumb = el}
                            tabIndex="0"
                        />
                    </Tooltip>
                </div>
                <Box className={css.maxLabel}>
                    {max}
                </Box>
            </Box>
        );
    }
}

export default Slider;
