// splunk-ui/src/components/Multiselect/examples/Fetching.jsx
import React, { Component } from 'react';
import { find, filter as filterArray } from 'lodash';
import filter from 'splunk-ui/util/filter';
import { _ } from 'splunk-ui/util/i18n';
import Multiselect from 'splunk-ui/components/Multiselect';
import { movies } from 'fixtures/movies.json';


class Example extends Component {
    constructor(props, ...rest) {
        super(props, ...rest);

        this.state = {
            values: [10, 30],
            keyword: '',
            isLoading: false,
        };
    }

    handleChange = (e, { values }) => {
        this.setState({ values, keyword: '' });
    }

    handleFilterChange = (e, { keyword }) => {
        this.setState({ keyword, isLoading: true });

        clearTimeout(this.timer);
        this.timer = setTimeout(this.fakeFinishFetch, 700);
    }

    fakeFinishFetch = () => {
        this.setState({ isLoading: false });
    }

    generateOptions = () => {
        let selectedOptions = [];

        // The selected items always have to be in the option list, but can be hidden
        if (this.state.values.length) {
            const isSelected = (movie) => !!find(this.state.values, (value) => movie.id === value);
            const selectedMovies = filterArray(movies, isSelected);
            selectedOptions = selectedMovies.map((movie, i) => (
                <Multiselect.Option
                    label={movie.title}
                    value={movie.id}
                    key={`selected-${i}`}
                    hidden
                />
            ));
        }

        if (this.state.isLoading) {
            // only return the select items
            return selectedOptions;
        }

        // simulate server-side filtering
        const filteredMovies = filter.filterByKeywords(
                movies,
                this.state.keyword,
                (movie) => movie.title
            );

        const list = filteredMovies.map((movie) => (
            <Multiselect.Option
                label={movie.title}
                value={movie.id}
                key={movie.id}
            />
        ));
        this.fullCount = list.length;

        // truncate list to simulate a fetch limit
        const listTruncated = list.slice(0, 20);
        this.count = listTruncated.length;

        return listTruncated.concat(selectedOptions);
    }

    footerMessage = () => {
        if (this.fullCount > this.count && !this.state.isLoading) {
            return _('%1 of %2 movies').replace('%1', this.count).replace('%2', this.fullCount);
        }
        return null;
    }

    render() {
        const options = this.generateOptions();
        const footerMessage = this.footerMessage();

        return (
            <Multiselect
                values={this.state.values}
                placeholder={_('Select a movie...')}
                onChange={this.handleChange}
                controlledFilter
                onFilterChange={this.handleFilterChange}
                isLoadingOptions={this.state.isLoading}
                footerMessage={footerMessage}
            >
                {options}
            </Multiselect>
        );
    }
}

export default Example;
