// splunk-ui/src/components/SlidingPanels/examples/Basic.jsx
import React, { Component } from 'react';
import Button from 'splunk-ui/components/Button';
import ControlGroup from 'splunk-ui/components/ControlGroup';
import Heading from 'splunk-ui/components/Heading';
import SlidingPanels from 'splunk-ui/components/SlidingPanels';
import P from 'splunk-ui/components/Paragraph';
import Text from 'splunk-ui/components/Text';

class Basic extends Component {

    constructor(...args) {
        super(...args);
        this.state = {
            activePanelId: 'one',
        };
        this.goBack = this.goBack.bind(this);
        this.goForward = this.goForward.bind(this);
    }

    goBack() {
        const nextPanel = this.state.activePanelId === 'one' ? 'two' : 'one';
        this.setState({
            activePanelId: nextPanel,
            transition: 'backward',
        });
    }

    goForward() {
        const nextPanel = this.state.activePanelId === 'one' ? 'two' : 'one';
        this.setState({
            activePanelId: nextPanel,
            transition: 'forward',
        });
    }

    render() {
        return (
            <div style={{ border: '1px solid #999' }}>
                <SlidingPanels {...this.state}>
                    <SlidingPanels.Panel
                        key="one"
                        panelId="one"
                        style={{ width: '100%', padding: 10 }}
                    >
                        <Heading>First</Heading>
                        <ControlGroup
                            label="Field Label"
                            tooltip="Tooltip helps explain the label."
                            help="This is the help text."
                        >
                            <Text canClear placeholder="optional" />
                        </ControlGroup>
                        <Button onClick={this.goBack}>Prev</Button>
                        <Button appearance="primary" onClick={this.goForward}>Next</Button>
                    </SlidingPanels.Panel>
                    <SlidingPanels.Panel
                        key="two"
                        panelId="two"
                        style={{ width: '100%', padding: 10 }}
                    >
                        <Heading>Second</Heading>
                        <P>
                            Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod
                            tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim
                            veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea
                            commodo consequat. Duis aute irure dolor in reprehenderit in voluptate
                            velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint
                            occaecat cupidatat non proident, sunt in culpa qui officia deserunt
                            mollit anim id est laborum.
                        </P>
                        <ControlGroup
                            label="Field Label"
                            tooltip="Tooltip helps explain the label."
                            help="This is the help text."
                        >
                            <Text canClear placeholder="optional" />
                        </ControlGroup>
                        <Button onClick={this.goBack}>Prev</Button>
                        <Button appearance="primary" onClick={this.goForward}>Next</Button>
                    </SlidingPanels.Panel>
                </SlidingPanels>
            </div>
        );
    }
}

export default Basic;
