// splunk-ui/src/components/SlidingPanels/examples/Modal.jsx
import React, { Component } from 'react';
import Button from 'splunk-ui/components/Button';
import ControlGroup from 'splunk-ui/components/ControlGroup';
import CollapsiblePanel from 'splunk-ui/components/CollapsiblePanel';
import Heading from 'splunk-ui/components/Heading';
import Modal from 'splunk-ui/components/Modal';
import SlidingPanels from 'splunk-ui/components/SlidingPanels';
import P from 'splunk-ui/components/Paragraph';
import Text from 'splunk-ui/components/Text';

const lorem = `
    Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do
    eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut
    enim ad minim veniam, quis nostrud exercitation ullamco laboris
    nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor
    in reprehenderit in voluptate velit esse cillum dolore eu fugiat
    nulla pariatur. Excepteur sint occaecat cupidatat non proident,
    sunt in culpa qui officia deserunt mollit anim id est laborum.
`;
class ModalExample extends Component {

    constructor(...args) {
        super(...args);
        this.state = {
            activePanelId: 'one',
            transition: 'forward',
            modalOpen: false,
        };
        this.goBack = this.goBack.bind(this);
        this.goForward = this.goForward.bind(this);
        this.open = this.open.bind(this);
        this.close = this.close.bind(this);
    }

    goBack() {
        const nextPanel = this.state.activePanelId === 'one' ? 'two' : 'one';
        this.setState({
            activePanelId: nextPanel,
            transition: 'backward',
        });
    }

    goForward() {
        const nextPanel = this.state.activePanelId === 'one' ? 'two' : 'one';
        this.setState({
            activePanelId: nextPanel,
            transition: 'forward',
        });
    }

    open() {
        this.setState({
            modalOpen: true,
        });
    }

    close() {
        this.setState({
            modalOpen: false,
        });
    }

    render() {
        const { activePanelId, modalOpen, transition } = this.state;
        return (
            <div>
                <Button label="Lorem ipsum" onClick={this.open} />
                <Modal onRequestClose={this.close} open={modalOpen}>
                    <Modal.Header title="SlidingPanels in a Modal" onRequestClose={this.close} />
                    <Modal.Body style={{ padding: 0 }}>
                        <SlidingPanels activePanelId={activePanelId} transition={transition}>
                            <SlidingPanels.Panel key="one" panelId="one" style={{ padding: 20 }}>
                                <Heading>First</Heading>
                                <ControlGroup
                                    label="Field Label"
                                    tooltip="Tooltip helps explain the label."
                                    help="This is the help text."
                                >
                                    <Text canClear placeholder="optional" />
                                </ControlGroup>
                            </SlidingPanels.Panel>
                            <SlidingPanels.Panel
                                key="two"
                                panelId="two"
                                style={{ padding: 20, width: '800px' }}
                            >
                                <CollapsiblePanel title="Sub Section">
                                    <P>{lorem}</P>
                                    <ControlGroup
                                        label="Field Label"
                                        tooltip="Tooltip helps explain the label."
                                        help="This is the help text."
                                    >
                                        <Text canClear placeholder="optional" />
                                    </ControlGroup>
                                </CollapsiblePanel>
                                <CollapsiblePanel title="Sub Section">
                                    <P>{lorem}</P>
                                </CollapsiblePanel>
                            </SlidingPanels.Panel>
                        </SlidingPanels>
                    </Modal.Body>
                    <Modal.Footer>
                        <Button onClick={this.goBack}>Prev</Button>
                        <Button appearance="primary" onClick={this.goForward}>Next</Button>
                    </Modal.Footer>
                </Modal>
            </div>
        );
    }
}

export default ModalExample;
