// splunk-ui/src/components/JSONTree/examples/Events.jsx
import React, { Component } from 'react';
import JSONTree from 'splunk-ui/components/JSONTree';
import Paragraph from 'splunk-ui/components/Paragraph';

class Basic extends Component {
    constructor(props) {
        super(props);
        this.state = {
            key: '',
            value: '',
        };
    }

    onClickJSONValue = (e, { key, value }) => {
        this.setState({ key, value });
    };

    render() {
        const obj = {
            glossary: {
                "title's": 'example glossary',
                GlossDiv: {
                    title: 'S',
                    GlossList: {
                        GlossEntry: {
                            ID: 'SGML',
                            SortAs: 'SGML',
                            GlossTerm: 'Standard Generalized Markup Language',
                            Acronym: 'SGML',
                            Abbrev: 'ISO 8879:1986',
                            GlossDef: {
                                para: 'Used to create markup languages such as DocBook.',
                                GlossSeeAlso: ['GML', 'XML', 1, true],
                            },
                            GlossSee: 'markup',
                        },
                    },
                },
            },
        };
        const { key, value } = this.state;
        return (
            <div>
                <JSONTree json={obj} onClickValue={this.onClickJSONValue} />
                <Paragraph>
                    Clicked Value: <span>{value}</span>
                    <br />
                    With key: <span>{key}</span>
                </Paragraph>
            </div>
        );
    }
}

export default Basic;
