// splunk-ui/src/components/Select/examples/Fetching.jsx
import React, { Component } from 'react';
import { find } from 'lodash';
import filter from 'splunk-ui/util/filter';
import { _ } from 'splunk-ui/util/i18n';
import Select from 'splunk-ui/components/Select';
import { movies } from 'fixtures/movies.json';


class Example extends Component {
    constructor(props, ...rest) {
        super(props, ...rest);

        this.state = {
            value: '',
            keyword: '',
            isLoading: false,
        };
    }

    handleChange = (e, { value }) => {
        this.setState({ value, keyword: '' });
    }

    handleFilterChange = (e, { keyword }) => {
        this.setState({ keyword, isLoading: true });

        clearTimeout(this.timer);
        this.timer = setTimeout(this.fakeFinishFetch, 700);
    }

    fakeFinishFetch = () => {
        this.setState({ isLoading: false });
    }

    generateOptions = () => {
        // The selected item always has to be in the option list, but can be hidden
        let selectedOption;
        if (this.state.value) {
            const selectedMovie = find(movies, (movie) => movie.id === this.state.value);

            selectedOption = (
                <Select.Option
                    label={selectedMovie.title}
                    value={selectedMovie.id}
                    key="selected"
                    hidden
                />
            );
        }

        if (this.state.isLoading) {
            // only return the selected item
            return selectedOption;
        }

        // simulate server-side filtering
        const filteredMovies = filter.filterByKeywords(
                movies,
                this.state.keyword,
                (movie) => movie.title
            );

        const list = filteredMovies.map((movie) => (
            <Select.Option
                label={movie.title}
                value={movie.id}
                key={movie.id}
            />
        ));
        this.fullCount = list.length;

        // truncate list to simulate a load limit
        const listTruncated = list.slice(0, 20);
        this.count = listTruncated.length;

        if (selectedOption) {
            listTruncated.push(selectedOption);
        }
        return listTruncated;
    }

    footerMessage = () => {
        if (this.fullCount > this.count && !this.state.isLoading) {
            return _('%1 of %2 movies').replace('%1', this.count).replace('%2', this.fullCount);
        }
        return null;
    }

    render() {
        const options = this.generateOptions();
        const footerMessage = this.footerMessage();

        return (
            <Select
                value={this.state.value}
                filter="controlled"
                placeholder={_('Select a movie...')}
                menuStyle={{ width: 300 }}
                onChange={this.handleChange}
                onFilterChange={this.handleFilterChange}
                isLoadingOptions={this.state.isLoading}
                footerMessage={footerMessage}
            >
                {options}
            </Select>
        );
    }
}

export default Example;
